Allow list_objects for anonymous users

**Is your feature request related to a problem? Please describe.**
We want to have an anonymously accessible (read-only) s3 bucket you can navigate through a webbrowser.
We noticed that while all necessary operations were allowed the call still failed with a 403.

`Aug-7 19:54:01.610 [Endpoint/13] [L0] core.endpoint.s3.s3_utils:: bucket_policy: action fit? s3:* s3:listbucket
Aug-7 19:54:01.610 [Endpoint/13] [L0] core.endpoint.s3.s3_utils:: bucket_policy: principal fit? SENSITIVE-7846cdd4c2b90527 undefined
Aug-7 19:54:01.610 [Endpoint/13] [L0] core.endpoint.s3.s3_utils:: bucket_policy: resource fit? /^arn:aws:s3:::website-foo-62044d78-39c5-4b65-9e0b-998350fda26c\/.*$/ arn:aws:s3:::website-foo-62044d78-39c5-4b65-9e0b-998350fda26c
Aug-7 19:54:01.610 [Endpoint/13] [L0] core.endpoint.s3.s3_utils:: bucket_policy: is_statements_fit true true false
Aug-7 19:54:01.610 [Endpoint/13] [ERROR] CONSOLE:: RPC._on_request: ERROR srv object_api.list_objects reqid 230@fcall://fcall(7pacxw7c) connid fcall://fcall(7pacxw7c) Error: not anonymous method list_objects
at RpcRequest.req.check_anonymous (/root/node_modules/noobaa-core/src/server/common_services/auth_server.js:528:44)
at authorize (/root/node_modules/noobaa-core/src/server/common_services/auth_server.js:436:17)
at RPC._on_request (/root/node_modules/noobaa-core/src/rpc/rpc.js:323:27)`

Troubleshooting has been done with @guymguym and he came to the conclusion that this call is not allowed for anonymous users.
[ref](https://github.com/noobaa/noobaa-core/blob/12b8a5beeb518d876f4fec22f12d714f57e47c17/src/api/object_api.js#L749)

**Describe the solution you'd like**
Since the policy allows it the call to list_objects should also be allowed.

**Describe alternatives you've considered**
In other cases we need to share either keys (directly or indirectly) and explain to users how to setup.
There is an existing s3 explorer project however it's not built right now for authentication and with auth we need to share keys with the app, change endpoint to use ODF instead of aws, ...
For each bucket we also need to create this setup again.

--- a/src/api/object_api.js
+++ b/src/api/object_api.js
@@ -746,7 +746,10 @@ module.exports = {
                     }
                 }
             },
-            auth: { system: ['admin', 'user'] }
+            auth: {
+                system: ['admin', 'user'],
+                anonymous: true,
+            }
         },
 
         list_object_versions: {

--- a/src/server/object_services/object_server.js
+++ b/src/server/object_services/object_server.js
@@ -695,6 +695,12 @@ async function read_object_mapping(req) {
     const { start, end, location_info } = req.rpc_params;
 
     const obj = await find_object_md(req);
+
+    // Check if the requesting account is authorized to read the object
+    if (!req.has_s3_bucket_permission(req.bucket, 's3:getobject', '/' + obj.key)) {
+        throw new RpcError('UNAUTHORIZED', 'requesting account is not authorized to read the object');
+    }
+
     const chunks = await map_reader.read_object_mapping(obj, start, end, location_info);
     const object_md = get_object_info(obj);
 
@@ -777,6 +783,12 @@ async function read_object_md(req) {
     }
 
     const obj = await find_object_md(req);
+
+    // Check if the requesting account is authorized to read the object
+    if (!req.has_s3_bucket_permission(req.bucket, 's3:getobject', '/' + obj.key)) {
+        throw new RpcError('UNAUTHORIZED', 'requesting account is not authorized to read the object');
+    }
+
     check_md_conditions(md_conditions, obj);
     const info = get_object_info(obj, { role: req.role });
     _check_encryption_permissions(obj.encryption, encryption);
@@ -1008,6 +1020,10 @@ async function delete_multiple_objects_by_filter(req) {
 async function list_objects(req) {
     dbg.log1('object_server.list_objects', req.rpc_params);
     load_bucket(req);
+
+    if (!req.has_s3_bucket_permission(req.bucket, "s3:listbucket")) {
+        throw new RpcError('UNAUTHORIZED', 'requesting account is not authorized to list objects');
+    }
 
     const limit = _list_limit(req.rpc_params.limit);
     if (!limit) return { is_truncated: false, objects: [], common_prefixes: [] };

--- a/src/test/unit_tests/coretest.js
+++ b/src/test/unit_tests/coretest.js
@@ -66,6 +66,7 @@ const rpc_client = server_rpc.rpc.new_client({
     auth_token: auth_server.make_auth_token({}),
     tracker: req => api_coverage.delete(req.srv),
 });
+const anon_rpc_client = server_rpc.rpc.new_client({});
 
 const SYSTEM = CORETEST;
 const EMAIL = `${CORETEST}@noobaa.com`;
@@ -665,6 +666,7 @@ exports.POOL_LIST = POOL_LIST;
 exports.PASSWORD = PASSWORD;
 exports.rpc_client = rpc_client;
 exports.new_rpc_client = new_rpc_client;
+exports.anon_rpc_client = anon_rpc_client;
 exports.get_http_address = get_http_address;
 exports.get_https_address = get_https_address;
 exports.get_https_address_sts = get_https_address_sts;

--- a/src/test/unit_tests/test_s3_bucket_policy.js
+++ b/src/test/unit_tests/test_s3_bucket_policy.js
@@ -1,9 +1,11 @@
 /* Copyright (C) 2016 NooBaa */
+/* eslint max-lines-per-function: ['error', 650] */
 'use strict';
 
 // setup coretest first to prepare the env
 const coretest = require('./coretest');
-const { rpc_client, EMAIL, POOL_LIST } = coretest;
+const { rpc_client, EMAIL, POOL_LIST, anon_rpc_client } = coretest;
+const MDStore = require('../../server/object_services/md_store').MDStore;
 coretest.setup({ pools_to_create: POOL_LIST });
 
 const AWS = require('aws-sdk');
@@ -29,6 +31,7 @@ const user_b = 'bob';
 let s3_a;
 let s3_b;
 let s3_owner;
+let s3_anon;
 
 const anon_access_policy = {
     Version: '2012-10-17',
@@ -39,6 +42,10 @@ const anon_access_policy = {
         Resource: [`arn:aws:s3:::*`]
     }]
 };
+
+// core_test_store is a global object created with the intention
+// to store data between tests.
+const cross_test_store = {};
 
 async function setup() {
     const self = this; // eslint-disable-line no-invalid-this
@@ -79,6 +86,11 @@ async function setup() {
     s3_creds.secretAccessKey = admin_keys[0].secret_key.unwrap();
     s3_owner = new AWS.S3(s3_creds);
     await s3_owner.createBucket({ Bucket: BKT }).promise();
+    s3_anon = new AWS.S3({
+        ...s3_creds,
+        accessKeyId: undefined,
+        secretAccessKey: undefined,
+    });
 }
 
 mocha.describe('s3_bucket_policy', function() {
@@ -457,5 +469,170 @@ mocha.describe('s3_bucket_policy', function() {
         await s3_owner.deleteBucketPolicy({ // should work - owner can always delete the buckets policy
             Bucket: BKT,
         }).promise();
+    });
+
+    mocha.it('anonymous user should be able to list bucket objects', async function() {
+        await s3_owner.putBucketPolicy({
+            Bucket: BKT,
+            Policy: JSON.stringify(anon_access_policy)
+        }).promise();
+
+        await s3_anon.makeUnauthenticatedRequest('listObjects', { Bucket: BKT }).promise();
+    });
+
+    mocha.it('anonymous user should not be able to list bucket objects when there is no policy', async function() {
+        await s3_owner.deleteBucketPolicy({
+            Bucket: BKT,
+        }).promise();
+
+        await assert_throws_async(s3_anon.makeUnauthenticatedRequest(
+            'listObjects',
+            { Bucket: BKT }
+        ).promise());
+    });
+
+    mocha.it('anonymous user should not be able to list bucket objects when policy doesn\'t allow', async function() {
+        const anon_deny_policy = {
+            Version: '2012-10-17',
+            Statement: [
+                {
+                    Effect: 'DENY',
+                    Principal: { AWS: "*" },
+                    Action: ['s3:GetObject', 's3:ListBucket'],
+                    Resource: [`arn:aws:s3:::*`]
+                },
+            ]
+        };
+
+        await s3_owner.putBucketPolicy({
+            Bucket: BKT,
+            Policy: JSON.stringify(anon_deny_policy)
+        }).promise();
+
+        await assert_throws_async(s3_anon.makeUnauthenticatedRequest(
+            'listObjects',
+            { Bucket: BKT }
+        ).promise());
+    });
+
+    mocha.it('[RPC TEST]: anonymous user should not be able to read_object_md when not explicitly allowed', async function() {
+        // Ensure that the bucket has no policy
+        await s3_owner.deleteBucketPolicy({
+            Bucket: BKT,
+        }).promise();
+
+        await assert_throws_async(anon_rpc_client.object.read_object_md({
+            bucket: BKT,
+            key: KEY,
+        }), 'requesting account is not authorized to read the object');
+    });
+
+    mocha.it('[RPC TEST]: anonymous user should be able to read_object_md when explicitly allowed', async function() {
+        const anon_read_policy = {
+            Version: '2012-10-17',
+            Statement: [
+                {
+                    Effect: 'Allow',
+                    Principal: { AWS: "*" },
+                    Action: ['s3:GetObject'],
+                    Resource: [`arn:aws:s3:::*`]
+                },
+            ]
+        };
+        await s3_owner.putBucketPolicy({
+            Bucket: BKT,
+            Policy: JSON.stringify(anon_read_policy)
+        }).promise();
+
+        cross_test_store.obj = await anon_rpc_client.object.read_object_md({
+            bucket: BKT,
+            key: KEY,
+        });
+    });
+
+    mocha.it('[RPC TEST]: anonymous user should be able to read_object_md when explicitly allowed to access only specific key', async function() {
+        const anon_read_policy_2 = {
+            Version: '2012-10-17',
+            Statement: [
+                {
+                    Effect: 'Allow',
+                    Principal: { AWS: "*"},
+                    Action: ['s3:GetObject'],
+                    Resource: [`arn:aws:s3:::${BKT}/${KEY}`]
+                },
+            ]
+        };
+        await s3_owner.putBucketPolicy({
+            Bucket: BKT,
+            Policy: JSON.stringify(anon_read_policy_2)
+        }).promise();
+
+        cross_test_store.obj = await anon_rpc_client.object.read_object_md({
+            bucket: BKT,
+            key: KEY,
+        });
+    });
+
+
+    mocha.it('[RPC TEST]: anonymous user should not be able to read_object_mapping when not explicitly allowed', async function() {
+        // Ensure that the bucket has no policy
+        await s3_owner.deleteBucketPolicy({
+            Bucket: BKT,
+        }).promise();
+
+        await assert_throws_async(anon_rpc_client.object.read_object_mapping({
+            bucket: BKT,
+            key: KEY,
+            obj_id: MDStore.instance().make_md_id(cross_test_store.obj.obj_id),
+        }), 'requesting account is not authorized to read the object');
+    });
+
+    mocha.it('[RPC TEST]: anonymous user should be able to read_object_mapping when explicitly allowed', async function() {
+        const anon_read_policy = {
+            Version: '2012-10-17',
+            Statement: [
+                {
+                    Effect: 'Allow',
+                    Principal: { AWS: "*" },
+                    Action: ['s3:GetObject'],
+                    Resource: [`arn:aws:s3:::*`]
+                },
+            ]
+        };
+        await s3_owner.putBucketPolicy({
+            Bucket: BKT,
+            Policy: JSON.stringify(anon_read_policy)
+        }).promise();
+
+        await anon_rpc_client.object.read_object_mapping({
+            bucket: BKT,
+            key: KEY,
+            obj_id: MDStore.instance().make_md_id(cross_test_store.obj.obj_id),
+        });
+    });
+
+    mocha.it('[RPC TEST]: anonymous user should be able to read_object_mapping when explicitly allowed to access only specific key', async function() {
+        const anon_read_policy_2 = {
+            Version: '2012-10-17',
+            Statement: [
+                {
+                    Effect: 'Allow',
+                    Principal: { AWS: "*"
+                    },
+                    Action: ['s3:GetObject'],
+                    Resource: [`arn:aws:s3:::${BKT}/${KEY}`]
+                },
+            ]
+        };
+        await s3_owner.putBucketPolicy({
+            Bucket: BKT,
+            Policy: JSON.stringify(anon_read_policy_2)
+        }).promise();
+
+        await anon_rpc_client.object.read_object_mapping({
+            bucket: BKT,
+            key: KEY,
+            obj_id: MDStore.instance().make_md_id(cross_test_store.obj.obj_id),
+        });
     });
 });

